hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/fgl/TestFineGrainedFSNamesystemLock.java

```diff
@@ -0,0 +1,335 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdfs.server.namenode.fgl;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.util.concurrent.HadoopExecutors;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class TestFineGrainedFSNamesystemLock {
+
+  private final Logger LOG = LoggerFactory.getLogger(TestFineGrainedFSNamesystemLock.class);
+
+  private int getLoopNumber() {
+    return ThreadLocalRandom.current().nextInt(2000, 3000);
+  }
+
+  @Test(timeout=120000)
+  public void testMultipleThreadsUsingLocks()
+      throws InterruptedException, ExecutionException {
+    FineGrainedFSNamesystemLock fsn = new FineGrainedFSNamesystemLock(new Configuration(), null);
+    ExecutorService service = HadoopExecutors.newFixedThreadPool(1000);
+
+    AtomicLong globalCount = new AtomicLong(0);
+    AtomicLong fsCount = new AtomicLong(0);
+    AtomicLong bmCount = new AtomicLong(0);
+    AtomicLong globalNumber = new AtomicLong(0);
+    AtomicLong fsNumber = new AtomicLong(0);
+    AtomicLong bmNumber = new AtomicLong(0);
+
+    List<Callable<Boolean>> callableList = new ArrayList<>(1000);
+    for (int i = 0; i < 1000; i++) {
+      int finalI = i;
+      int index = finalI % 12;
+      if (index == 0) {
```

Comment:
Can add comments for each case. For example, test for the global write lock here?